modules/imgproc/src/canny.cpp

```diff
@@ -62,60 +62,77 @@ static void CannyImpl(Mat& dx_, Mat& dy_, Mat& _dst, double low_thresh, double h
 
 
 #ifdef HAVE_IPP
-static bool ippCanny(const Mat& _src, Mat& _dst, float low,  float high)
+template <bool useCustomDeriv>
+static bool ippCanny(const Mat& _src, const Mat& dx_, const Mat& dy_, Mat& _dst, float low, float high)
```

Comment:
we can add an internal assert that either `_src` or both `dx_` and `dy_` must be empty.